dbsnap_verify/__init__.py

```diff
@@ -0,0 +1,154 @@
+    else:
+        set_state_doc_in_s3(state_doc, "wait")
+        logger.info("Did not find the {snapshot_date} snapshot of {database}".format(**state_doc))
+        logger.info("Going to sleep.")
+
+def restore(state_doc):
+    description = get_database_descripton(state_doc)
+    if description is None:
+        set_state_doc_in_s3(state_doc, "restore")
+        start_restore_from_snapshot(state_doc)
+    elif description[DBInstanceStatus] == "available":
+        set_state_doc_in_s3("verify")
+        verify(state_doc, description)
+        
+def verify(state_doc, description):
+    reset_master_password()
+    connection = connect_to_endpoint(description["endpoint"])
+    result = run_all_the_tests(connection, state_doc["verfication_checks"])
+    if result:
+        set_state_doc_in_s3("success")
+        alarm(state_doc, "success")
+    else:
+        set_state_doc_in_s3("alarm")
+        alarm(state_doc, "error")
+    set_state_doc_in_s3("cleanup")
+    clean_up(state_doc)
+
+def cleanup(state_doc):
+    description = get_database_descripton(state_doc)
+    if description is None:
+        # start waiting for tomorrows date.
+        set_state_doc_in_s3("wait")
+    elif description[DBInstanceStatus] == "available":
+        destroy_database(state_doc)
+
+def alarm(state_doc):
+    # trigger an alarm, maybe cloudwatch or something.
+    pass
+
+def upload_state_doc(state_doc):
+    state_doc_json = json.dumps(state_doc)
+    s3.put_object(
+        Bucket=state_doc["state_dot_bucket"],
+        Key=state_doc_s3_key(event["database"]),
+        Body=state_doc_json,
+    )
+
+def set_state_doc_in_s3(state_doc, new_state):
+    if "states" not in state_doc:
+        state_doc["states"] = []
+    state_doc["states"].append(
+        {
+            "state" : new_state,
+            "timestamp" : now_timestamp(),
+        }
+    )
+    upload_state_doc(state_doc)
+    return state_doc
+
+def state_doc_s3_key(database):
+    return "state-doc-{}.json".format(database)
+
+def download_state_doc(event):
+    s3_object = s3.get_object(
+        Bucket=event["state_doc_bucket"],
+        Key=state_doc_s3_key(event["database"]),
+    )
+    # download state_doc json from s3 and stick it into a string.
+    state_doc_json = s3_object["Body"].read()
+    # turn json into a dict and return it.
+    return json.loads(state_doc_json)
+
+def get_or_create_state_doc_in_s3(event):
+    """get (or create if missing) the state_doc in S3."""
+    try:
+        state_doc = download_state_doc(event)
+    except s3.exceptions.NoSuchKey:
+        state_doc = event
+        state_doc["snapshot_date"] = datetime_to_date_str(tomorrow_date())
+        state_doc = set_state_doc_in_s3(state_doc, "wait")
+    return state_doc
+
+def current_state(state_doc):
+    return state_doc["states"][-1]["state"]
+
+state_handlers = {
+  "wait": wait,
+  "restore": restore,
+  "verify": verify,
+  "cleanup": cleanup,
+  "alarm": alarm,
+}
+
+def lambda_handler(event, context):
```

Comment:
I don't believe this can be in a top level module :(

Reply:
Oh good catch, I'm brand new to lambda so I wasn't sure about the needed structure.  I think we can put a dummy `lambda_handler` in the root if that is true which calls this one.

Reply:
Yeah, it's a pain - you have to specify what the function is in lambda, and for some reason I don't think you can do `.lambda_handler` you have to specify a submodule in the package.  Itmight be that i just haven't figured out the right syntax though.